### PR TITLE
Switch retry

### LINKED ISSUE #12672: Look into switching ExponentialRetry (base_client) to azure-core RetryPolicy
pipeline error with hosts

### PR DESCRIPTION
**Pull request by seankane-msft**
Changes the storage retry policy to inherit from the `azure.core.pipelines.policies.RetryPolicy`.
Fixes #12672

### CODE REVIEW

File: sdk/tables/azure-data-tables/azure/data/tables/_policies.py [hunk 1]
```diff
@@ -353,19 +354,64 @@ def on_response(self, request, response):
...
+        self,
+        initial_backoff=15, # type: int
+        increment_base=3, # type: int
+        retry_total=10, # type: int
+        retry_to_secondary=False, # type: bool
+        random_jitter_range=3, # type: int
+        **kwargs # type: Any
+    ):
+        """
+        Build a BLANK retry object.
```
Comment (iscai-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/_policies.py:371]: nit: build a StorageRetryPolicy

File: sdk/tables/azure-data-tables/azure/data/tables/_policies.py [hunk 2]
```diff
@@ -353,19 +354,64 @@ def on_response(self, request, response):
...
 
-    def __init__(self, **kwargs):
-        self.total_retries = kwargs.pop('retry_total', 10)
+        :param int initial_backoff:
+            The initial backoff interval, in seconds, for the first retry.
+        :param int increment_base:
+            The base, in seconds, to increment the initial_backoff by after the
+            first retry.
+        :param int max_attempts:
```
Comment (iscai-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/_policies.py:378]: not seeing `max_attepmts`in the init for this policy. Does this mean this sohuld be a keyword param, or is it something we're missing

File: sdk/tables/azure-data-tables/azure/data/tables/_policies.py [hunk 3]
```diff
@@ -353,19 +354,64 @@ def on_response(self, request, response):
...
+        """
+        self.initial_backoff = initial_backoff
+        self.increment_base = increment_base
+        self.random_jitter_range = random_jitter_range
+        self.total_retries = retry_total
         self.connect_retries = kwargs.pop('retry_connect', 3)
         self.read_retries = kwargs.pop('retry_read', 3)
         self.status_retries = kwargs.pop('retry_status', 3)
-        self.retry_to_secondary = kwargs.pop('retry_to_secondary', False)
+        self.retry_to_secondary = retry_to_secondary
```
Comment (iscai-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/_policies.py:396]: I think we should at least pass the kwargs to the parent `RetryPolicy`, Doesn't really look like there's much of an overlap though between the parameters fro `StorageRetryPolicy` and `RetryPoliccy`

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_policies_async.py [hunk 4]
```diff
@@ -112,7 +155,7 @@ async def send(self, request):  # pylint: disable =W0236
                 break
             except AzureError as err:
                 retries_remaining = self.increment(
-                    retry_settings, request=request.http_request, error=err)
+                    retry_settings, error=err)
```
Comment (iscai-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/aio/_policies_async.py:158]: just a question: why do you no longer need to pass request?

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_policies_async.py [hunk 5]
```diff
@@ -161,7 +204,7 @@ def __init__(self, initial_backoff=15, increment_base=3, retry_total=3,
         super(ExponentialRetry, self).__init__(
             retry_total=retry_total, retry_to_secondary=retry_to_secondary, **kwargs)
 
-    def get_backoff_time(self, settings, **kwargs):
+    def get_backoff_time(self, settings, **kwargs): # pylint: disable=unused-argument
```
Comment (iscai-msft) [inline sdk/tables/azure-data-tables/azure/data/tables/aio/_policies_async.py:207]: if kwargs is unused, you can remove it like you did for the sync `get_backoff_time` and other methods

File: sdk/tables/azure-data-tables/azure/data/tables/_policies.py (hunk 1 above)
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/azure/data/tables/_policies.py:371]: Thank you for the catch!

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_policies_async.py (hunk 4 above)
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/azure/data/tables/aio/_policies_async.py:158]: `request` isn't used in `self.increment` at all, pylint noticed that for me

File: sdk/tables/azure-data-tables/azure/data/tables/_policies.py (hunk 2 above)
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/azure/data/tables/_policies.py:378]: `max_attempts` and `retry_total` are synonymous in this situation, removing `max_attempts`

File: sdk/tables/azure-data-tables/azure/data/tables/_base_client.py [hunk 6]
```diff
@@ -50,7 +50,8 @@
     StorageRequestHook,
     StorageResponseHook,
     StorageLoggingPolicy,
-    StorageHosts, ExponentialRetry,
+    StorageHosts,
+    StorageRetryPolicy,
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/azure/data/tables/_base_client.py:54]: Can we renamed to TablesRetryPolicy?

File: sdk/tables/azure-data-tables/azure/data/tables/_policies.py [hunk 7]
```diff
@@ -608,7 +642,7 @@ def __init__(self, backoff=15, retry_total=3, retry_to_secondary=False, random_j
         super(LinearRetry, self).__init__(
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/azure/data/tables/_policies.py:642]: This is not longer the super class

Comment (annatisch) [reply sdk/tables/azure-data-tables/azure/data/tables/_policies.py:642]: Wait - never mind - misunderstood the diff here...

**Pull request merged**